transaction-logger/src/main.rs

```diff
@@ -0,0 +1,731 @@
+struct AccountAddressEq(AccountAddress);
+
+impl From<AccountAddressEq> for AccountAddress {
+    fn from(aae: AccountAddressEq) -> Self { aae.0 }
+}
+
+impl PartialEq for AccountAddressEq {
+    fn eq(&self, other: &Self) -> bool {
+        let bytes_1: &[u8; 32] = &self.0.as_ref();
+        let bytes_2: &[u8; 32] = &other.0.as_ref();
+        bytes_1[0..29] == bytes_2[0..29]
+    }
+}
+
+impl Hash for AccountAddressEq {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let bytes: &[u8; 32] = self.0.as_ref();
+        bytes[0..29].hash(state);
+    }
+}
+
+impl AsRef<AccountAddressEq> for AccountAddress {
+    fn as_ref(&self) -> &AccountAddressEq { unsafe { std::mem::transmute(self) } }
+}
+
+#[derive(Debug, Error)]
+enum NodeError {
+    /// Error establishing connection.
+    #[error("Error connecting to the node {0}.")]
+    ConnectionError(tonic::transport::Error),
+    /// Query errors, etc.
+    #[error("Error querying the node {0}.")]
+    OtherError(#[from] anyhow::Error),
+}
+
+/// Return Err if querying the node failed.
+/// Return Ok(()) if the channel to the database was closed.
+async fn use_node(
+    node_ep: endpoints::Endpoint,
+    token: &str,
+    sender: &tokio::sync::mpsc::Sender<TransactionLogData>,
+    height: &mut AbsoluteBlockHeight, // start height
+    max_parallel: u32,
+    stop_flag: &AtomicBool,
+    canonical_cache: &mut HashSet<AccountAddressEq>,
+) -> Result<(), NodeError> {
+    let mut node = endpoints::Client::connect(node_ep, token.into())
+        .await
+        .map_err(NodeError::ConnectionError)?;
+    // if the cache is empty we seed it with all accounts at the last finalized
+    // block. This will only happen the first time this function is successfully
+    // invoked.
+    if canonical_cache.is_empty() {
+        let info = node
+            .get_consensus_status()
+            .await
+            .context("Error querying consensus status.")?;
+        let accounts = node
+            .get_account_list(&info.last_finalized_block)
+            .await
+            .context("Error querying account list.")?;
+        // this relies on the fact that get_account_list returns canonical addresses.
+        log::debug!(
+            "Initializing the address cache with {} accounts.",
+            accounts.len()
+        );
+        for addr in accounts {
+            canonical_cache.insert(AccountAddressEq(addr));
+        }
+    }
+    let mut parallel = 1;
+    while !stop_flag.load(Ordering::Acquire) {
+        let mut handles = Vec::with_capacity(parallel as usize);
+        for h in u64::from(*height)..u64::from(*height) + parallel {
+            let mut node = node.clone();
+            handles.push(async move {
+                let height: AbsoluteBlockHeight = h.into();
+                let blocks_at_height = node.get_blocks_at_height(height.into()).await?;
+                if blocks_at_height.len() != 1 {
+                    return Ok(Err(QueryFailure::NoBlocks));
```

Comment:
Could be more than one block?

Reply:
Yes, if we passed last finalized height.